keps/sig-api-machinery/20190507-kube-controller-manager-config.md

```diff
@@ -0,0 +1,537 @@
+versioned configuration files. It can consume a beta-versioned config file and many flags are now deprecated and
+pending removal, in favor of this file. Many remaining flags will be replaced by the file over time. Additionally,
+the kube-proxy component is very close to having a beta-versioned config file of its own. We have do some versioned
+componentConfig about controller manager, we must make it work finally.
+In this KEP, we aims to add controller manager config file support, by which we can configure controller manager via
+a config file instead of command line flags. Only in this way, we can get rid of ourselves from the complex arguments of
+command line.
+
+## Motivation
+
+### Goals
+
+- Sync with others components, such as kube-proxy, kubelet, kube-scheduler, which support consume a config file.
+- No requirement to restart a process when you change a file, unlike flags.
+- Configure controller manager via a config file instead of command line flags.
+- Enable dynamic configuration deployment mechanisms.
+
+### Non-Goals
+
+- flags are nonstandard interfaces with weak stability guarantees. They are confusing and hard to deploy, and this
+  is the opposite of what Kubernetes should be.
+- put forward the work about upgrade versioned component.
+
+## Proposal
+
+This proposal contains three logical units of work. Each subsection is explained in more detail as below.
+- Step 1: Group related options together
+- Step 2: Using "option+config" pattern
+  we have done some basically work about `Step 1` and `Step 2` before, such as
+  - [controller-manager: switch to options+config pattern and add https+auth](https://github.com/kubernetes/kubernetes/pull/59582) by [@sttts](https://github.com/sttts)
+  - [split up the huge set of flags into smaller option structs](https://github.com/kubernetes/kubernetes/pull/60270) by [@stewart-yu](https://github.com/stewart-yu)
+  - [split the generic component config and options into a kube and cloud part](https://github.com/kubernetes/kubernetes/pull/63283) by [@stewart-yu](https://github.com/stewart-yu)
+  - [move specific option sub-struct from controller-manager into kube-controller manager packages](https://github.com/kubernetes/kubernetes/pull/64142) by [@stewart-yu](https://github.com/stewart-yu)
+  - [Implement the --controllers flag fully for the cloud-controller manager](https://github.com/kubernetes/kubernetes/pull/68283) by [@stewart-yu](https://github.com/stewart-yu)
+
+- Step 3: Enable consume a config file instead of command line flags.
+ - [Implement a dedicated serializer package for ComponentConfigs](https://github.com/kubernetes/kubernetes/pull/74111) by [@stewart-yu](https://github.com/stewart-yu)
+ - [Moving ComponentConfig API types to staging repos](https://github.com/kubernetes/community/blob/master/keps/sig-cluster-lifecycle/0014-20180707-componentconfig-api-types-to-staging.md) by [@luxas](https://github.com/luxas) and [@sttts](https://github.com/sttts)
```

Comment:
The above link has changed to https://github.com/kubernetes/enhancements/blob/master/keps/sig-cluster-lifecycle/wgs/0014-20180707-componentconfig-api-types-to-staging.md